wip/beam-spec.md

```diff
@@ -0,0 +1,437 @@
+execution of the function that was preempted by the interrupt handler.)
+
+- BEAM's interrupt handling capabilities can be summarized as: prioritization of
+  interrupts, individual interrupt masking and global interrupt masking.
+
+- The BEAM has 8 *interrupts* whose *handlers* are named: `INTERRUPT0`,
+  `INTERRUPT1`, all the way to `INTERRUPT7`.
+
+- The signature of all interrupt handlers must be `[unsafe] fn()`. That is these
+  handlers don't take arguments nor can they return values.
+
+- If not defined in the program, interrupt handlers default to:
+
+``` rust
+#[no_mangle]
+unsafe fn INTERRUPT_() {
+    loop {}
+}
+```
+
+- When the BEAM starts, all interrupts are *disabled*, *unmasked* and are
+  configured with an initial *priority* of `1`; and the *running priority* of
+  the system is set to `0`.
+
+- Interrupts can be assigned any *priority* in the range `1..=255`.
+
+- At any point in time an interrupt signal may arrive and change the state of an
+  interrupt to *pending*. That is interrupt signals occur asynchronously to the
+  execution of the program.
+
+- If interrupts are *enabled*, setting the state of an *unmasked* interrupt to
+  *pending* results in its handler preempting the current function *if and only
+  if* the *priority* of the interrupt is *higher* than the current *running
+  priority*, otherwise the interrupt will be kept in the *pending* state until
+  the *running priority* is low enough to allow preemption.
+
+- The *pending* state of an interrupt gets cleared when its handler starts
+  executing.
+
+- Once an interrupt handler starts executing the *running priority* is
+  automatically set to the *priority* of the interrupt being serviced. When the
+  interrupt handler returns the running priority is restored to its previous
+  (before preemption) value.
+
+- While interrupts are *disabled* preemption can not occur. Interrupts in the
+  pending state will remain in that state until interrupts are re-enabled.
+
+- Setting the state of a *masked* interrupt to *pending* will never result in
+  preemption, no matter the priority of the interrupt. Such interrupt will only
+  be serviced after it has been unmasked.
+
+- The behavior of setting the *running priority* within the context of an
+  interrupt handler to a value *lower* than the *priority* of that interrupt
+  results in UNSPECIFIED behavior.
+
+- Changing the *priority* of an interrupt from within its handler results in
+  UNSPECIFIED behavior.
+
+- Interrupt handling can be summarized as follows:
+
+``` rust
+// pseudo-code
+if interrupts_are_enabled {
+    if interrupt_N_is_pending {
+        if !interrupt_N_is_masked {
+            if interrupt_N_priority > running_priority {
+                atomic {
+                    interrupt_N_is_pending = false;
+                    running_priority = interrupt_N_priority;
+                    start_INTERRUPTN();
+                }
+            }
+        }
+    }
+}
```

Comment:
This would have been really useful further up when I tried to figure out what all that English text means.